File: csharp/Dialects/SqliteDialect.cs
using Microsoft.Data.Sqlite;

namespace Deterministic.MigrateRunner;

internal static class SqliteDialect
{
    public static readonly ISqlDialect Instance = new SqlDialect(
        "sqlite",
        ["SQLITE_PATH", "DB_PATH"],
        useTransaction: true,
        NormalizeConnection,
        s => new SqliteConnection(s),
        (cmd, n, v) => SqlDialect.Bind(cmd, $"${n}", v),
        ident => $"\"{ident}\"",
        name => $"${name}",
        connection =>
        {
            var path = FilesystemPath(connection);
            return path is null || File.Exists(path)
                ? null
                : HelpTemplates.Message(
                    "errors/sqlite-prerequisite",
                    new Dictionary<string, string> { ["path"] = path }
                );
        }
    );

    private static string NormalizeConnection(string connection)
    {
        var c = connection ?? string.Empty;
        if (c.Contains('=', StringComparison.Ordinal))
            return c;
        if (c.StartsWith("sqlite://", StringComparison.OrdinalIgnoreCase))
        {
            var path = c["sqlite://".Length..];
            return string.IsNullOrEmpty(path) ? "Data Source=:memory:" : $"Data Source={path}";
        }
        return string.IsNullOrEmpty(c) ? "Data Source=:memory:" : $"Data Source={c}";
    }

    private static string? FilesystemPath(string? connection)
    {
        var s = (connection ?? string.Empty).Trim();
        var eq = s.IndexOf('=', StringComparison.Ordinal);
        if (eq >= 0)
        {
            var key = s[..eq].Trim().Replace(" ", string.Empty);
            if (string.Equals(key, "DataSource", StringComparison.OrdinalIgnoreCase))
            {
                s = s[(eq + 1)..].Trim();
                var semi = s.IndexOf(';', StringComparison.Ordinal);
                if (semi >= 0)
                    s = s[..semi].Trim();
            }
        }
        foreach (var prefix in new[] { "sqlite://", "sqlite:", "file:" })
        {
            if (s.StartsWith(prefix, StringComparison.OrdinalIgnoreCase))
            {
                s = s[prefix.Length..];
                break;
            }
        }
        return s.Length == 0 || string.Equals(s, ":memory:", StringComparison.Ordinal) ? null : s;
    }
}

File: typescript/src/dialects/sqlite-dialect.ts
import { mkdir } from "node:fs/promises";
import { dirname } from "node:path";
import type {
  MigrationClient,
  MigrationRow,
  MigrationSqlParam,
} from "../migration-client.ts";
import { pathExists } from "../path-exists.ts";
import { SqlDialectBase } from "./sql-dialect-base.ts";

export const sqliteFilesystemPath = (connection: string): string | null => {
  let s = connection.trim();
  const eq = s.indexOf("=");
  if (eq >= 0 && /data\s*source/i.test(s.slice(0, eq))) {
    s = s.slice(eq + 1).trim();
  } else if (/^sqlite:\/\//i.test(s)) {
    s = s.slice("sqlite://".length);
  } else if (/^sqlite:/i.test(s)) {
    s = s.slice("sqlite:".length);
  } else if (/^file:/i.test(s)) {
    s = s.slice("file:".length);
  }
  if (s === ":memory:" || s === "") return null;
  return s;
};

export class SqliteDialect extends SqlDialectBase {
  readonly name = "sqlite" as const;
  readonly connectionEnvironmentVariables = ["SQLITE_PATH", "DB_PATH"] as const;
  override readonly usesLastInsertRowid = true;

  async prerequisiteError(connection: string): Promise<string | null> {
    const path = sqliteFilesystemPath(connection);
    if (path === null) return null;
    if (await pathExists(path)) return null;
    return `sqlite file: ${path} does not exist — run 'migrate-setup --provider sqlite --connection ${path}' to create it`;
  }

  async prepareSetup(connection: string): Promise<void> {
    const filePath = sqliteFilesystemPath(connection);
    if (filePath !== null) {
      await mkdir(dirname(filePath), { recursive: true });
    }
  }

  async createClient(connection: string): Promise<MigrationClient> {
    if (!connection) throw new Error("sqlite requires a database file path");
    const { default: Database } = await import("better-sqlite3");
    const db = new Database(connection);
    return {
      dialect: "sqlite",
      async exec(sql) {
        db.exec(sql);
      },
      async query(sql, params = []) {
        const stmt = db.prepare<MigrationSqlParam[], MigrationRow>(sql);
        try {
          return stmt.all(...params);
        } catch {
          stmt.run(...params);
          return [];
        }
      },
      async run(sql, params = []) {
        db.prepare(sql).run(...params);
      },
      async transaction(fn) {
        db.exec("BEGIN");
        try {
          const r = await fn();
          db.exec("COMMIT");
          return r;
        } catch (e) {
          try {
            db.exec("ROLLBACK");
          } catch (rollbackErr) {
            console.warn(
              "sqlite ROLLBACK failed after transaction error; original error rethrown",
              rollbackErr,
            );
          }
          throw e;
        }
      },
      async close() {
        db.close();
      },
    };
  }
}
